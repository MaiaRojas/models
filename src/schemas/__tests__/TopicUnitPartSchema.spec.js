const mongoose = require('mongoose/browser');
const { JSDOM } = require('jsdom');
const { TopicUnitSchema, TopicUnitPartSchema } = require('../')(mongoose, (new JSDOM()).window.document);
const babelTopicJson = require('./fixtures/topics/babel');

describe('TopicUnitPartSchema', () => {
  it('should fail validation when fields missing', () => {
    const doc = new mongoose.Document({}, TopicUnitPartSchema);
    expect(doc.validateSync().errors).toMatchSnapshot();
  });

  it('should fail validation when type read and missing body', () => {
    const doc = new mongoose.Document({
      slug: '00-opening',
      duration: 5,
      durationString: '5min',
      format: 'self-paced',
      type: 'read',
      title: 'Opening',
    }, TopicUnitPartSchema);

    return doc.validate()
      .catch(err => expect(err.message).toBe('Body is required for part type read'));
  });

  it('should pass validation when type quiz and missing body', () => {
    const unit = new mongoose.Document({}, TopicUnitSchema);
    const doc = new mongoose.Document({
      unit: unit._id,
      slug: '01-quiz',
      duration: 10,
      durationString: '10min',
      format: 'self-paced',
      type: 'quiz',
      title: 'Quiz 1',
    }, TopicUnitPartSchema);

    return doc.validate();
  });

  it('should pass validation when type practice and missing body', () => {
    const unit = new mongoose.Document({}, TopicUnitSchema);
    const doc = new mongoose.Document({
      unit: unit._id,
      slug: '02-exercises',
      duration: 10,
      durationString: '10min',
      format: 'self-paced',
      type: 'practice',
      title: 'Exercises',
    }, TopicUnitPartSchema);

    return doc.validate();
  });

  it('should pass validation when type practice and empty body', () => {
    const unit = new mongoose.Document({}, TopicUnitSchema);
    const doc = new mongoose.Document({
      unit: unit._id,
      slug: '02-exercises',
      duration: 10,
      durationString: '10min',
      format: 'self-paced',
      type: 'practice',
      title: 'Exercises',
      body: '',
    }, TopicUnitPartSchema);

    return doc.validate();
  });

  it('should pass validation when type read and has body', () => {
    const body = 'Blah <h1>bla<span>h.*</span>-</h1>+_ blah...';
    const unit = new mongoose.Document({}, TopicUnitSchema);
    const doc = new mongoose.Document({
      unit: unit._id,
      slug: '00-opening',
      duration: 5,
      durationString: '5min',
      format: 'self-paced',
      type: 'read',
      title: 'Opening',
      body,
    }, TopicUnitPartSchema);
    expect(doc.validateSync()).toBe(undefined);
    expect(doc.body).toBe(body);
    expect(doc.searchableBody).toBe('Blah blah.*-+_ blah...');
  });

  it('should validate existing topic json', () => {
    const unit = new mongoose.Document({}, TopicUnitSchema);
    const parts = Object.values(babelTopicJson.syllabus).reduce(
      (memo, unitJson) => memo.concat(
        Object.keys(unitJson.parts).map(slug => ({
          unit: unit._id,
          slug,
          ...unitJson.parts[slug],
        })),
      ),
      [],
    );
    return Promise.all(parts.map((partJson) => {
      const part = new mongoose.Document(partJson, TopicUnitPartSchema);
      return part.validate();
    }));
  });
});
