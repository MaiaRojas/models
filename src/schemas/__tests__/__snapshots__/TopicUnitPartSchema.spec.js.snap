// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopicUnitPartSchema should fail validation when fields missing 1`] = `
Object {
  "duration": [ValidatorError: Path \`duration\` is required.],
  "durationString": [ValidatorError: Path \`durationString\` is required.],
  "format": [ValidatorError: Path \`format\` is required.],
  "slug": [ValidatorError: Path \`slug\` is required.],
  "title": [ValidatorError: Path \`title\` is required.],
  "type": [ValidatorError: Path \`type\` is required.],
  "unit": [ValidatorError: Path \`unit\` is required.],
}
`;
